InteractionRegions are missing for some `display: table` elements

etiennesegonzac opened the pull request:
#### fde1853a7a1574c5468741b2022d4d9d6ae1cf24
<pre>
InteractionRegions are missing for some `display: table` elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=257379">https://bugs.webkit.org/show_bug.cgi?id=257379</a>
&lt;rdar://109538261&gt;

Reviewed by Tim Horton.

Don&apos;t skip the child renderer&apos;s InteractionRegion if the link parent is
`display: table`, like we do for inline links.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
Check if the matched renderer `isDisplayTableOrTablePart()` before
bailing out.

* LayoutTests/interaction-region/display-table-expected.txt: Added.
* LayoutTests/interaction-region/display-table.html: Added.
Add a test.

* LayoutTests/interaction-region/layer-tree-expected.txt:
Test update.

Canonical link: <a href="https://commits.webkit.org/264598@main">https://commits.webkit.org/264598@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b04ffac7092f82b1587842cf0e5f5788b6535adb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8035 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8320 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9687 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8139 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8039 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10306 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8231 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11004 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8179 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9266 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7332 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9817 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6567 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7352 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14941 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7690 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7476 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10846 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7947 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6468 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7261 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/7265 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1938 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11469 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7688 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -258,7 +258,7 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
 
     // The parent will get its own InteractionRegion.
-    if (!isOriginalMatch && !isInlineNonBlock)
+    if (!isOriginalMatch && !isInlineNonBlock && !renderer.style().isDisplayTableOrTablePart())
```
I think on the web, some people use `display: table` or similar to do clearfix hacks, not sure that affects the code you're writing, but worth noting.

etiennesegonzac replied:
No impact here, but happy to know one of the reasons for this setup!

etiennesegonzac committed:
InteractionRegions are missing for some `display: table` elements
https://bugs.webkit.org/show_bug.cgi?id=257379
<rdar://109538261>

Reviewed by Tim Horton.

Don't skip the child renderer's InteractionRegion if the link parent is
`display: table`, like we do for inline links.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
Check if the matched renderer `isDisplayTableOrTablePart()` before
bailing out.

* LayoutTests/interaction-region/display-table-expected.txt: Added.
* LayoutTests/interaction-region/display-table.html: Added.
Add a test.

* LayoutTests/interaction-region/layer-tree-expected.txt:
Test update.

Canonical link: https://commits.webkit.org/264598@main
